src/listpack.c

```diff
@@ -972,6 +979,76 @@ unsigned char *lpDelete(unsigned char *lp, unsigned char *p, unsigned char **new
     return lpInsert(lp,NULL,NULL,0,p,LP_REPLACE,newp);
 }
 
+/* Delete a range of entries from the listpack start with the element pointed by 'p'. */
+unsigned char *lpDeleteRangeWithEntry(unsigned char *lp, unsigned char **p, unsigned long num) {
+    size_t bytes = lpBytes(lp);
+    unsigned long deleted = 0;
+    unsigned char *eofptr = lp + bytes - 1;
+    unsigned char *first, *tail;
+    first = tail = *p;
+
+    if (num == 0) return lp;  /* Nothing to delete, return ASAP. */
+
+    /* Find the next entry to the last entry that needs to be deleted.
+     * lpLength may be unreliable due to corrupt data, so we cannot
+     * treat 'num' as the number of elements to be deleted. */
+    while (num--) {
+        deleted++;
+        tail = lpSkip(tail);
+        if (tail[0] == LP_EOF) break;
+        lpAssertValidEntry(lp, bytes, tail);
+    }
+
+    /* Store the offset of the element 'first', so that we can obtain its
+     * address again after a reallocation. */
+    unsigned long poff = first-lp;
+
+    /* Move tail to the front of the ziplist */
+    memmove(first, tail, eofptr - tail + 1);
+    lpSetTotalBytes(lp, bytes - (tail - first));
+    uint32_t numele = lpGetNumElements(lp);
+    if (numele != LP_HDR_NUMELE_UNKNOWN)
+        lpSetNumElements(lp, numele-deleted);
+    lp = lpShrinkToFit(lp);
+
+    /* Store the entry. */
+    *p = lp+poff;
+    if ((*p)[0] == LP_EOF) *p = NULL;
+
+    return lp;
+}
+
+/* Delete a range of entries from the listpack. */
+unsigned char *lpDeleteRange(unsigned char *lp, long index, unsigned long num) {
+    unsigned char *p;
+    uint32_t numele = lpGetNumElements(lp);
+
+    if (num == 0) return lp; /* Nothing to delete, return ASAP. */
+    if ((p = lpSeek(lp, index)) == NULL) return lp;
+
+    if (numele == LP_HDR_NUMELE_UNKNOWN) {
+        /* If the listpack length cannot be obtained in constant time,
+         * using lpDeleteRangeWithEntry will be much faster. */
+        lp = lpDeleteRangeWithEntry(lp, &p, num);
```

Comment:
i don't understand this. it looks like the alternative is doing the same (also using lpDeleteRangeWithEntry)

Reply:
Yes, but if we do that, we can't use lpGetNumElements, we need to use lpLength, which will traverse the entire listpack when listpack length > UINT16_MAX. Instead, we use lpDeleteRangeWithEntry to delete the length, so it only traverses to the index.

Reply:
ok, i tihnk i see what you mean.
the code below has an "optimization" to just move the EOF marker, and that optimization can't be used without knowing the real length (which in this case would require calling `lpLength`).
we rather not use `lpLength`, so instead in this case we skip the EOF "optimization" and just fall back to the normal path.

the current code and comment is confusing because it mentions we can do something faster, but the thing it's faster from, isn't at all there..
i suggest to refactor that code and improve the comment.

it can be something like:
```
/* If we know we're gonna delete beyond the end of the listpack, we can just move
 * the EOF marker, and there's no need to iterate though the entries.
 * but if we can't be sure how many entries there are, we rather avoid calling lpLength
 * since that means an additional iteration on all elements. */
if (we know the real length, and we know we're deleting beyond the range) {
    do the optimal thing and move the EOF
} else {
    lpDeleteRangeWithEntry
}
````

Reply:
Good.

Reply:
Done.